File: source/go/bfs_shortest.go
package main

import (
	"fmt"
	"strings"
)

func bfsShortestPath(adj [][]int, start, end, n int) (int, []int) {
	visited := make([]bool, n)
	dist := make([]int, n)
	prev := make([]int, n)
	for i := range dist {
		dist[i] = -1
		prev[i] = -1
	}

	queue := []int{start}
	visited[start] = true
	dist[start] = 0

	for len(queue) > 0 {
		u := queue[0]
		queue = queue[1:]
		if u == end {
			break
		}
		for _, v := range adj[u] {
			if !visited[v] {
				visited[v] = true
				dist[v] = dist[u] + 1
				prev[v] = u
				queue = append(queue, v)
			}
		}
	}

	var path []int
	current := end
	for current != -1 {
		path = append([]int{current}, path...)
		current = prev[current]
	}

	return dist[end], path
}

func main() {
	adj := [][]int{
		{1, 2},
		{3},
		{3, 4},
		{5},
		{5},
		{},
	}

	distance, path := bfsShortestPath(adj, 0, 5, 6)
	fmt.Printf("distance: %d\n", distance)
	parts := make([]string, len(path))
	for i, v := range path {
		parts[i] = fmt.Sprintf("%d", v)
	}
	fmt.Printf("path: %s\n", strings.Join(parts, " "))
}

File: source/javascript/bfs_shortest.js
// BFS Shortest Path (Unweighted)
function bfsShort(adj, start, end, n) {
    const visited = new Array(n).fill(false);
    const dist = new Array(n).fill(-1);
    const prev = new Array(n).fill(-1);

    const queue = [start];
    visited[start] = true;
    dist[start] = 0;

    while (queue.length > 0) {
        const u = queue.shift();
        if (u === end) break;
        for (const v of adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                dist[v] = dist[u] + 1;
                prev[v] = u;
                queue.push(v);
            }
        }
    }

    const path = [];
    let current = end;
    while (current !== -1) {
        path.unshift(current);
        current = prev[current];
    }

    return { distance: dist[end], path };
}

const adj = [[1, 2], [3], [3, 4], [5], [5], []];
const result = bfsShort(adj, 0, 5, 6);
console.log(`distance: ${result.distance}`);
console.log(`path: ${result.path.join(" ")}`);
